### PR TITLE
Improve logging message for filestore CWL runs

### PR DESCRIPTION
**Pull request by chapmanb**
The output logging messages when running cwltoil pipelines were generic
references to jobstore directories, making it hard for users to
determine which stage of the pipeline was running. This fixes in two
ways:

- Moves references to internal file deletion to debug, instead of info.
- Handles the case where we're running CWL-based runs with filesystem
  storage, reading the underlying CWL and providing a useful logging
  message from the baseCommand.

This is an attempt to improve the logging output during cwltoil runs to make it easier for users to identify what stage of the pipeline they were at. In the previous iteration, users could tell jobs were running but had little insight into the current step since all the outputs are references to `_toil_worker` commands, internal jobstore IDs and cleanup of files:
```
Executing command: '_toil_worker /home/chapmanb/drive/work/cwl/test_bcbio_cwl/cwltoil_work/cwltoil_jobstore l/1/jobt5ZNIZ'.
Got message from job at time 07-05-2016 10:34:15: Successfully deleted local copies of file with ID 'k/K/tmpABzeZz.tmp'.
Got message from job at time 07-05-2016 10:34:15: Successfully deleted local copies of file with ID 'T/u/tmpC8PsBX.tmp'.
Executing command: '_toil_worker /home/chapmanb/drive/work/cwl/test_bcbio_cwl/cwltoil_work/cwltoil_jobstore v/i/jobxQaXXO'.
```
After the change, the output looks like:
```
Executing command: 'bcbio_nextgen.py runfn prep_samples cwl'.
Executing command: 'bcbio_nextgen.py runfn prep_align_inputs cwl'.
Jobstore directory is: /home/chapmanb/drive/work/cwl/test_bcbio_cwl/cwltoil_work/cwltoil_jobstore
Executing command: '_toil_worker file:/home/chapmanb/drive/work/cwl/test_bcbio_cwl/cwltoil_work/cwltoil_jobstore K/2/jobVgjucx'.
Executing command: '_toil_worker file:/home/chapmanb/drive/work/cwl/test_bcbio_cwl/cwltoil_work/cwltoil_jobstore 6/n/job2ND5wI'.
Executing command: 'bcbio_nextgen.py runfn process_alignment cwl'.
Executing command: 'bcbio_nextgen.py runfn process_alignment cwl'.
```
I'm happy for suggestions to make this better and more generalized. Right now it would only work for filesystem-based JobStores and CWL runs. I was not able to understand the abstractions to know the best way to get these commands for all cases so it will fall back to the previous method where it is not these cases. You can see that behavior in the new example above where it falls back to the old method when executing non CWLTool commands (like workflows and scatter operations). Thanks much for looking at this.

### CODE REVIEW

**Comment by benedictpaten** [comment]
Jenkins, test this please.

**Comment by hannes-ucsc** [comment]
We shouldn't have to parse the worker command line to get at that information. That's just is brittle and circuitous. This will be solved permanently and allencompassingly by #652 and #677. However, I am torn as to whether we should include this anyways as a temporary patch for a real paint point for our users.

Would like to get @benedictpaten and @fnothaft position on this.

**Comment by fnothaft** [comment]
> However, I am torn as to whether we should include this anyways as a temporary patch for a real paint point for our users.

What's the timeline for a fix to #652 and #677? My thought is that since the pain point is real and big, unless the all-encompassing fix is expected to hit very soon (e.g. within the next release), I would merge this in, and then revert it when we've got an all-encompassing fix.

**Comment by hannes-ucsc** [comment]
> What's the timeline for a fix to #652 and #677?

I guessing, three months.

**Comment by hannes-ucsc** [comment]
@arkal there are some subtle changes to how caching is logged. Could you take a look? That one indention change worries me most.

**Comment by hannes-ucsc** [comment]
@chapmanb, we use camelCase in Toil. Could you rename _get_debug_cmd accordingly? And, pending @arkal's feedback, please rebase with conflict resolution.

**Comment by arkal** [comment]
Uhm... deleteGlobal and deleteLocal are actually user callable functions too. In that case we definitely want to see the logs in INFO and not in DEBUG. An option is to change the signature of `deleteGlobalFile` and `deleteLocalFile` to become `(self, fileStoreID, verbose=False)` and then modify [this line](https://github.com/chapmanb/toil/blob/967fc1f79792592164a9ec39d1ad223fa8571c22/src/toil/job.py#L2291) to `fileStore.deleteGlobalFile(jobStoreFileID, quiet=True)`
quiet = True will put those 2 logs into DEBUG, quiet=False will keep then in INFO

File: src/toil/job.py [hunk 1]
```diff
@@ -982,9 +982,9 @@ def deleteLocalFile(self, fileStoreID):
                             os.remove(cachedFile)
                             cacheInfo.cached -= fileSize
                     self.logToMaster('Successfully deleted cached copy of file with ID '
-                                     '\'%s\'.' % fileStoreID)
+                                     '\'%s\'.' % fileStoreID, logging.DEBUG)
                 self.logToMaster('Successfully deleted local copies of file with ID '
-                                 '\'%s\'.' % fileStoreID)
+                                 '\'%s\'.' % fileStoreID, logging.DEBUG)
```
Comment (benedictpaten) [inline src/toil/job.py]: I agree with this change, I would go further and turn these into standard log messages. @arkal - in the case that the job succeeds do you really want to see this message? Seems like a lot of low level information. The current design for "global" logging messages is to convey them all back to the leader. If you want to persist this behavior we should filter log to masters on the worker by the log level to prevent excessive log passing during a production run.

File: src/toil/job.py [hunk 2]
```diff
@@ -1013,7 +1013,7 @@ def deleteGlobalFile(self, fileStoreID):
             # Add the file to the list of files to be deleted once the run method completes.
             self.filesToDelete.add(fileStoreID)
             self.logToMaster('Added file with ID \'%s\' to the list of files to be' % fileStoreID +
-                             ' globally deleted.')
+                             ' globally deleted.', logging.DEBUG)
```
Comment (benedictpaten) [inline src/toil/job.py]: Ditto comment above on logging. @arkal

**Comment by benedictpaten** [comment]
I left comments on the log messages. I agree that sorting out the log messages is important. What we have now is the accumulated cruft of our development cycles. Totally agree with Hannes about this being a pain point.

I really like the attempt to get at the "job name", but I agree with Hannes, I don't like that this is a single machine only fix. I had forgotten that single machine had those log messages. For only a modest amount of storage cost we could afford to serialise the job name in the job wrapper and pass it around, and then all log messages could access it. We could make the job name a class variable is users wanted a nicer string (not the limitations of the class name), or even just a variable of the class if a user wanted to identify a particular instance of a class. The biggest cost of this approach would be an increase of memory usage on the leader.

**Comment by arkal** [comment]
Yes I agree they should be standard logging. I think i made those logToMaster because at the time I thought that was replacing standard logging.

@benedictpaten i think if we see "Downloaded file from file store " and "wrote file to file store" for all files read and written by the user, we should also see "deleted file from file store" messages. In the interest of symmetricity

**Comment by brainstorm** [comment]
That pull request would be really useful since I'm now trying to implement support for SLURM flags in CWLToil (https://github.com/BD2KGenomics/toil/issues/918) but running into some issues with some code in `leader.py` and `cwltoil.py` when it tries to unpickle the `rootJobReturnValue`.

Will pull it in my branch, but happy to see it in `master`. Good job @chapmanb!

**Comment by chapmanb** [comment]
Hannes, Arjun, Benedict and Roman;
Thanks for the thoughts. I'm happy to try some of the suggestions and improvements in this patch if there is consensus we should move forward. I'd love to have the less brittle fix command serialization suggested by Benedict but don't have enough Toil code understanding to implement it myself.

In defense of this patch I can give you the CWL user perspective. The current set of log messsages are completely meaningless to me. I can tell that something is running but have no idea about where in the process it is. I don't care at all about Toil staging and deleting files unless it's somehow erroring out, and the jobStore outputs have no useful information to act on. I feel most strongly that we need something to indicate status that is meaningful to the user who has no concept of the underlying CWL or code, so happy to get there with whatever path you think is best.

File: src/toil/job.py (hunk 1 above)
Comment (hannes-ucsc) [inline src/toil/job.py]: +1 on changing these to standard logging.

@chapmanb, why did you change this? Depending on your response, we might ask you take undo these two changes out and we might address them in a second PR.

**Comment by hannes-ucsc** [comment]
We will merge this once the sticking points are resolved. The code will be removed when we have a more complete solution.

Also see my line comments.

File: src/toil/job.py (hunk 1 above)
Comment (hannes-ucsc) [inline src/toil/job.py]: Just fixed a bunch of typos in my previous line note.

Comment (chapmanb) [inline src/toil/job.py]: I switched these from the default (logging.INFO) to DEBUG as they fill up the logging input and don't offer much helpful information from the user perspective. For my CWL runs I get hundreds of messages like:
```
Got message from job at time 07-13-2016 10:15:21: Successfully deleted local copies of file with ID 'h/y/tmpMJXEcB.tmp'.
Got message from job at time 07-13-2016 10:15:21: Successfully deleted local copies of file with ID 'j/W/tmp2cwAzR.tmp'.
Got message from job at time 07-13-2016 10:15:21: Successfully deleted local copies of file with ID 'G/2/tmpGT1jJU.tmp'.
Got message from job at time 07-13-2016 10:15:21: Successfully deleted local copies of file with ID 'a/W/tmpUWRh4C.tmp'.
Got message from job at time 07-13-2016 10:15:21: Successfully deleted local copies of file with ID 'A/l/tmpwS225a.tmp'.
Got message from job at time 07-13-2016 10:15:21: Successfully deleted local copies of file with ID 'Q/f/tmpMt3I3j.tmp'.
```
In my mind these seemed better suited for debugging than running information.

**Comment by chapmanb** [comment]
Thanks, I left a note. My logging changes are meant to adjust messages from INFO to DEBUG to avoid some noise in the output. Let me know if I can provide more information.

File: src/toil/job.py (hunk 1 above)
Comment (arkal) [inline src/toil/job.py]: are you using job.fileStore.deleteLocalFile?

Comment (hannes-ucsc) [inline src/toil/job.py]: CWL might be triggering the calls to deleteLocalFile.

Comment (chapmanb) [inline src/toil/job.py]: That's right on. We're calling this through CWL so are a layer separated from using the Toil API.

Comment (hannes-ucsc) [inline src/toil/job.py]: It seem like there is a theme developing: we want invocations of file store methods made by Toil code to be logged at DEBUG, while user script invocations should be logged at INFO. We have @arkal's `quiet=` kwarg proposal. We could also use `inspect` to figure out whether the calling stack frame is for a function belonging to Toil. Other ideas?

Comment (arkal) [inline src/toil/job.py]: Make it a command line flag?

Comment (arkal) [inline src/toil/job.py]: Also i realised, the messages you're seeing are from the job cleanup calls to deleteLocalFile. I've made a comment on how you can fix this inline

File: src/toil/job.py [hunk 3]
```diff
@@ -982,9 +982,9 @@ def deleteLocalFile(self, fileStoreID):
                             os.remove(cachedFile)
                             cacheInfo.cached -= fileSize
                     self.logToMaster('Successfully deleted cached copy of file with ID '
-                                     '\'%s\'.' % fileStoreID)
+                                     '\'%s\'.' % fileStoreID, logging.DEBUG)
                 self.logToMaster('Successfully deleted local copies of file with ID '
```
Comment (arkal) [inline src/toil/job.py]: If this statement is moved into the if block, it will be printed only when the user explicitly calls job.fileStore.deleteLocalFile.

Comment (arkal) [inline src/toil/job.py]: I can address this in a future PR

File: src/toil/job.py (hunk 1 above)
Comment (hannes-ucsc) [inline src/toil/job.py]: Ok, it's clear that we need a different ticket to track this. @chapmanb can you remove that change from this PR and create a GH issue for it?

**Comment by chapmanb** [comment]
Arjun and Hannes -- thanks for the comments and discussion. I removed the logging messages from this PR and opened up an issue to track that (#1061). You both have a much better clue than I do about what has to happen so I'll stay away from a PR on that.

Hope this works for getting the debugging merged. Thanks again.

File: src/toil/batchSystems/singleMachine.py [hunk 4]
```diff
@@ -120,6 +120,33 @@ def __init__(self, config, maxCores, maxMemory, maxDisk):
             self.workerThreads.append(worker)
             worker.start()
 
+    def _get_debug_cmd(self, jobCommand):
```
Comment (hannes-ucsc) [inline src/toil/batchSystems/singleMachine.py]: Toil uses camel case. Please rename to _getDebugCmd.

Comment (chapmanb) [inline src/toil/batchSystems/singleMachine.py]: Thanks Hannes, fixed. Apologies, it's quite hard to get used to camelCase when you normally use the pep8 styles.

**Comment by hannes-ucsc** [comment]
Please squash.

File: src/toil/batchSystems/singleMachine.py (hunk 4 above)
Comment (hannes-ucsc) [inline src/toil/batchSystems/singleMachine.py]: I dislike camelCase in Python. But I dislike inconsistency even more.

On Thu, Jul 14, 2016 at 10:18 AM, Brad Chapman <notifications@github.com>
wrote:

> In src/toil/batchSystems/singleMachine.py
> <https://github.com/BD2KGenomics/toil/pull/1048#discussion_r70845448>:
>
> > @@ -120,6 +120,33 @@ def __init__(self, config, maxCores, maxMemory, maxDisk):
> >              self.workerThreads.append(worker)
> >              worker.start()
> >
> > +    def _get_debug_cmd(self, jobCommand):
>
> Thanks Hannes, fixed. Apologies, it's quite hard to get used to camelCase
> when you normally use the pep8 styles.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/BD2KGenomics/toil/pull/1048/files/d71a123c4a03b56ef322cdeb0270a847bf83f6bc#r70845448>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AE562MZN5BXY41uAREp6dfnJeLRzShJoks5qVm9QgaJpZM4JG9ej>
> .
>



-- 
Hannes Schmidt
Software Engineer
Center for Biomolecular Science and Engineering
University of California, Santa Cruz

(206) 696-2316 (cell)
hannes@ucsc.edu

**Comment by chapmanb** [comment]
Also squashed

**Comment by hannes-ucsc** [comment]
Will merge once build passes.

**Comment by hannes-ucsc** [comment]
Jenkins, test this please.
